ops/relation.py

```diff
@@ -0,0 +1,548 @@
+    to be compatible then the :class:`Consumer` object raises a
+    :class:`ProviderAvailable` event to inform the consumer charm, that
+    a relation with the provider charm has been successful.
+    """
+    def __init__(self, handle, data=None):
+        super().__init__(handle)
+        self.data = data
+
+    def snapshot(self):
+        return {"data": self.data}
+
+    def restore(self, snapshot):
+        self.data = snapshot["data"]
+
+
+class ProviderInvalid(EventBase):
+    """Event triggered when a provider is not compatible
+
+    When a consumer charm forms a relation with a provider charm,
+    their :class:`Consumer` and :class:`Provider` object exchange
+    information to asertain compatibility. If they relation is found
+    not to be compatible then the :class:`Consumer` object raises a
+    :class:`ProviderInvalid` event to inform the consumer charm, that
+    a relation with the provider charm has *not* been successful.
+    """
+    def __init__(self, handle, data=None):
+        super().__init__(handle)
+        self.data = data
+
+    def snapshot(self):
+        return {"data": self.data}
+
+    def restore(self, snapshot):
+        self.data = snapshot["data"]
+
+
+class ProviderUnready(EventBase):
+    """Event triggered when a provider is not ready
+
+    If a provider charm is not ready to service requests, when a
+    consumer charm for a new relation with it, or is already related
+    to it, then a :class:`ProviderUnready` event is raised. This
+    presumes that the provider charm as set its `ready` status to
+    `False`.
+
+    The :class:`ProviderUnready` event is raised regardless of wether
+    the provider charm is compatible or not. Compatibility checks are
+    done only if the provider charm is ready to service requests.
+    """
+    pass
+
+
+class ProviderBroken(EventBase):
+    """Event raised when provider consumer relation is disolved
+
+    If the realtion between a provider and consumer charm is removed,
+    then a :class:`ProviderBroken` relation is raised.
+    """
+    pass
+
+
+class ConsumerEvents(CharmEvents):
+    """Descriptor for consumer charm events
+    """
+    available = EventSource(ProviderAvailable)
+    invalid = EventSource(ProviderInvalid)
+    unready = EventSource(ProviderUnready)
+    broken = EventSource(ProviderBroken)
+
+
+class Consumer(Object):
+    """Manages relations with a resource provider
+
+    The :class:`Consumer` object manages relations with resource
+    provider charms, by checking compatibility between consumer
+    requirements and provider type and version specification. Any
+    charm that uses resources provided by other charms may manage its
+    relation with the providers by instantiating on :class:`Consumer`
+    object for each such relation. A :class:`Consumer` object may be
+    instantiated in the `__init__` method of the consumer charm as
+    follows
+
+    Example::
+
+        self.provider_name = Consumer(self, resource_name, consumes)
```

Comment:
you use resource_name here, but isn't this 'relation_name'?

Reply:
Ops. Yes. Thank you. Will fix.

Reply:
Fixed.